repository_service_tuf/helpers/tuf.py

```diff
@@ -180,15 +178,21 @@ def _update_snapshot(
             f"signing threshold '{threshold}'"
         )
 
-        root.roles[role_name] = Role([], threshold)
-        for signer in signers:
-            root.add_key(
-                Key.from_securesystemslib_key(signer.key_dict), role_name
-            )
+        roles[role_name] = Role([], threshold)
+        keys: list[tuple[Key, str]] = [
+            (Key.from_securesystemslib_key(signer.key_dict), role_name)
+            for signer in signers
+        ]
+
+    root = Root(roles=roles)
+    for key in keys:
+        root.add_key(key[0], key[1])
 
     # Add signature wrapper, bump expiration, and sign and persist
     for role in [targets, snapshot, timestamp, root]:
-        metadata = Metadata(role)
+        # TODO: Create an issue for this mypy error if it's not resolved by the
+        # PR reviews
+        metadata = Metadata(role)  # type: ignore
```

Comment:
Really not sure why you have `mypy` issue for this.
Strange.
@lukpueh any idea why we have that issue?
Asking you as this seems to be connected to `python-tuf` `Generic[T]` type.

Reply:
Looks like we need to annotate the list used in the for loop `[targets, snapshot, timestamp, root]`...

Reply:
Hm. I know what the problem is, but not how to fix it. :) Btw. error message is:
```
repository_service_tuf/helpers/tuf.py:195: error: Value of type variable "T" of "Metadata" cannot be "Signed"  [type-var]
```

So it looks like the typing system sees a list of different objects with a common base class and probably annotates it as something like `List[Signed]`. But Metadata requires its first argument to be not just any `Signed`, but one of the currently available `Signed` subclasses aka `T`.

My naive approach would be to declare `role: T` or  `roles: List[T]`, but then mypy complains about `T` being unbound.

🤷

Reply:
> it looks like the typing system sees a list of different objects with a common base class and probably annotates it as something like List[Signed]

yes, you can confirm this by adding `reveal_type(role)` inside the loop (that will only work during mypy)

I don't think there's a clean way to handle this but some ideas:
* could annotate as `role: Union[Root, ...]` but that likely just makes mypy not do anything with it -- the mypy ignore is likely just as useful
* instead of having e.g. a `root=Root()` as the modified object initialize `root_md: Metadata(Root())` instead in the beginning (and then access root_md.signed if you need to -- it's annotated so works just fine)

I would go for the second option. Yes, it's annoying to have to change the flow you had in mind for annotations but I think that style usually works just as well and you get working annotations.

Reply:
oh that's interesting. I suppose the for loop mis-typing still breaks it...

> Do you know why we can't just do role: T?

T is not a "real type" IIRC: it doesn't exist at runtime in anyway and so can only be used in a function signature (and then _maybe_ within that function, I don't recall the details).

In any case, I suggest trying  `Metadata(Root())`.

Reply:
> In any case, I suggest trying Metadata(Root()).

Agree, that looks like the best solution.

Wanted to say it's really nice to see you both in a discussion here.
Felt like a discussion on a `python-tuf` contribution for a moment. 😁

Reply:
Thanks for the very useful input everyone, ` Metadata(Root())` seems to be doing the trick!